MM-20895 Experimental Sidebar Features - Custom Category Look n Feel

devinbinnie opened the pull request:
#### Summary
This PR contains the modals and components needs to support custom categories.
Currently you can use the + menu to add a new category and have it show up in the sidebar, but it will not allow to add channels.
This is here to demo the look and feel.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-20895

esethna commented:
@devinbinnie can you help list what is still in progress on this PR UX is aware and isn't reporting known issues?

devinbinnie commented:
> AWESOME! Few notes based on the spec:
> 
> 1. Create category modal should not allow user input beyond 22 characters (same code as rename channel modal)
> 2. We should probably truncate the category names if they are long
>    ![image](https://user-images.githubusercontent.com/14333569/80034528-6bb85680-84a3-11ea-899d-e641cc744425.png)
> 3. We need to implement the rules for removing the drop zone helper and NEW badge
>    ![image](https://user-images.githubusercontent.com/14333569/80034675-a4583000-84a3-11ea-982e-88161d7619e7.png)
> 
> Approving from PM perspective, @andrewbrown00 please help with UX review
> 
> 1. @andrewbrown00 should we allow or add an emoji picker? Perhaps a follow-up ticket?

@esethna 1 and 2 are fixed. 3 will be implemented once this PR is merged with the Drag n Drop PR.

jgilliam17 commented:
Thank you @devinbinnie 
Looks good, one question - is shadow intended?
<img width="242" alt="Screen Shot 2020-04-23 at 2 04 43 PM" src="https://user-images.githubusercontent.com/52937121/80133985-13995700-856c-11ea-93fb-a4144923c487.png">

devinbinnie commented:
> Thank you @devinbinnie
> Looks good, one question - is shadow intended?
> <img alt="Screen Shot 2020-04-23 at 2 04 43 PM" width="242" src="https://user-images.githubusercontent.com/52937121/80133985-13995700-856c-11ea-93fb-a4144923c487.png">

Yes it is AFAIK. @andrewbrown00 can confirm.

andrewbrown00 commented:
> > Thank you @devinbinnie
> > Looks good, one question - is shadow intended?
> > <img alt="Screen Shot 2020-04-23 at 2 04 43 PM" width="242" src="https://user-images.githubusercontent.com/52937121/80133985-13995700-856c-11ea-93fb-a4144923c487.png">
> 
> Yes it is AFAIK. @andrewbrown00 can confirm.

@jgilliam17 the shadow is not intended. @devinbinnie did you see it in the design file?

![image](https://user-images.githubusercontent.com/48559658/80153767-27ed4c00-858c-11ea-9ceb-e4b8875e14ae.png)

devinbinnie commented:
> > > Thank you @devinbinnie
> > > Looks good, one question - is shadow intended?
> > > <img alt="Screen Shot 2020-04-23 at 2 04 43 PM" width="242" src="https://user-images.githubusercontent.com/52937121/80133985-13995700-856c-11ea-93fb-a4144923c487.png">
> > 
> > 
> > Yes it is AFAIK. @andrewbrown00 can confirm.
> 
> @jgilliam17 the shadow is not intended. @devinbinnie did you see it in the design file?
> 
> ![image](https://user-images.githubusercontent.com/48559658/80153767-27ed4c00-858c-11ea-9ceb-e4b8875e14ae.png)

It doesn't show up visually, but there's a `box-shadow` in the generated CSS:

```
position: absolute;
width: 208px;
height: 40px;
left: 16px;
top: 32px;

border: 1px dashed rgba(255, 255, 255, 0.4);
box-sizing: border-box;
/* Elevation 4 */

box-shadow: 0px 8px 24px rgba(0, 0, 0, 0.12);
border-radius: 4px;
```

devinbinnie commented:
> Looking awesome @devinbinnie 🎉
> 
> **Few items to resolve**
> 
> * [x]  Focused state on the input field is using the wrong style, should use `var(--button-bg)`. This was my fault, I've updated the Figma file.
> 
> Current	Actual
> ![image](https://user-images.githubusercontent.com/48559658/80154786-4eac8200-858e-11ea-8372-1c43c7406109.png)	![image](https://user-images.githubusercontent.com/48559658/80154825-58ce8080-858e-11ea-917d-dbc61e45e14c.png)

Fixed!

> * [x]  Help text should align to the left edge of the input field
> 
> ![image](https://user-images.githubusercontent.com/48559658/80155897-a51ac000-8590-11ea-9cad-d840565141e6.png)
> 

Fixed!

> * [x]  Disabled state of the create button should use the following opacities:
>   `color: var(--center-channel-color-64);`
>   `background: var(--center-channel-color-08);`
> 
> Current	Actual
> ![image](https://user-images.githubusercontent.com/48559658/80155146-18bbcd80-858f-11ea-82c8-6c2966db1c46.png)	![image](https://user-images.githubusercontent.com/48559658/80155289-6d5f4880-858f-11ea-82cb-aa14c42941e4.png)

Fixed!

> * [ ]  Input field in the create category modal needs to support the max character rule of 64; i.e. don't support inputting more than 64 characters and also trimming characters if a user pastes in a string with more than 64 characters.
> 
> ![image](https://user-images.githubusercontent.com/48559658/80154169-0b054880-858d-11ea-9721-7177d72fd252.png)
> 
> * [ ]  Make sure the we have space between the input area and the text string
> 
> Current	Expected
> ![image](https://user-images.githubusercontent.com/48559658/80154394-7cdd9200-858d-11ea-86e9-ca0961a13c88.png)	![image](https://user-images.githubusercontent.com/48559658/80154366-68999500-858d-11ea-9368-f4d6de29c68c.png)

The above 2 shouldn't be necessary as the spec only allows for 22 character long category names, so I've set the input to only allow for 22 characters.

> * [x]  Already commented about the shadow needing to be removed from the new category placeholder. Also, I provided the `hand` icon and it's in the update `mattermost icons` font to match the design.
> 
> Current	Expected
> ![image](https://user-images.githubusercontent.com/48559658/80155504-d5ae2a00-858f-11ea-8cca-d2be1fd15d01.png)	![image](https://user-images.githubusercontent.com/48559658/80155539-e5c60980-858f-11ea-8c43-1d5fc17f5980.png)

Fixed!

> * [x]  Already commented on, but we need to truncate category header to match today rules; we'll also need to account for the extra space of the `NEW` badge when it's present.
> 
> ![image](https://user-images.githubusercontent.com/48559658/80156118-22decb80-8591-11ea-862e-ba7678033833.png)

Fixed!

> 
> * [x]  The **dropzone** UI should hide when collapsing a category header, [spec doc](https://docs.google.com/document/d/17Uw_HMptsXADgganXffjB1k8_WuAn5nyOU5JrqZj_AA/edit#)
> 
> Current	Expected
> ![image](https://user-images.githubusercontent.com/48559658/80156251-76e9b000-8591-11ea-9468-f1191261b160.png)	![image](https://user-images.githubusercontent.com/48559658/80156282-88cb5300-8591-11ea-80e1-7c26e896f5f9.png)

Fixed!

> * [ ]  Hovering over a newly created category header should replace the `NEW` badge with the `...` icon
> 
> ![image](https://user-images.githubusercontent.com/48559658/80156536-2d4d9500-8592-11ea-9dbf-28c4c8812d78.png)

This will come once the branches have been merged together.

esethna commented:
@devinbinnie the test server seems to be having issues ^

devinbinnie commented:
@esethna I think the Cloud team might be doing maintenance or something, I've noticed a few PRs having this issue.

andrewbrown00 commented:
@devinbinnie any updates on what's going on with the `test server`?

devinbinnie commented:
> @devinbinnie any updates on what's going on with the `test server`?

devinbinnie commented:
> > @devinbinnie any updates on what's going on with the `test server`?

Sent a message to Cloud Team, they're going to look into it.

cpanato commented:
@devinbinnie @andrewbrown00 issue fixed, sorry for the trouble that caused

devinbinnie commented:
> @devinbinnie @andrewbrown00 issue fixed, sorry for the trouble that caused

No worries, thanks for fixing!

andrewbrown00 commented:
@devinbinnie @esethna created this follow up ticket to increase the opacity of just the emoji in the category name https://mattermost.atlassian.net/browse/MM-24634

esethna commented:
@deanwhillier @hmhealey ready for dev review

andrewbrown00 commented:
Also [created an improvement ticket](https://mattermost.atlassian.net/browse/MM-24635) for the character limit

deanwhillier commented:
Initially reviewing on my iPad Pro:

- Should the `Drag channels here...` box be collapsible? I’m thinking not. @andrewbrown00, thoughts?
- Drag & Drop doesn’t seem to work on touchscreen, so the `Drag channels here...` box is misleading. Is it possible with React DnD to support DnD on touchscreens (i.e. am I doing something wrong)?

devinbinnie commented:
> Initially reviewing on my iPad Pro:
> 
> * Should the `Drag channels here...` box be collapsible? I’m thinking not. @andrewbrown00, thoughts?
> * Drag & Drop doesn’t seem to work on touchscreen, so the `Drag channels here...` box is misleading. Is it possible with React DnD to support DnD on touchscreens (i.e. am I doing something wrong)?

It should be working on mobile, it works on my phone at the very least. I could test on the simulator but it probably wouldn't be the same thing...

deanwhillier commented on sass/components/_categories.scss:
```diff
@@ -0,0 +1,140 @@
+            padding: 0;
+
+            .close {
+                color: v(center-channel-color-56);
```
Is`v()` an scss mixin/function for `--var()`?

devinbinnie replied:
Not sure, this was copy/pasted from the channel switcher modal.
But I would wager yes.

deanwhillier replied:
Given the potential confusion, can we please change this for the proper --var()? It's only an extra 4 characters and then removes the need for a mixin/function. I realize it's used elsewhere and I will suggest the same in those cases if we refactor/rebuild.

deanwhillier commented on sass/layout/_sidebar-left.scss:
```diff
@@ -335,6 +335,61 @@
+        }
+
+        &.collapsed {
+            height: 0 !important;
```
I'm assuming the `!important` is absolutely necessary here? If so, I'm wondering if we should maybe have. a comment to explain why it is being used?

devinbinnie replied:
Yeah I can add a comment.

devinbinnie commented on components/edit_category_modal.tsx:
```diff
@@ -0,0 +1,117 @@
+                        </h1>
+                    </div>
+                    <div className='edit-category__body'>
+                        <QuickInput
+                            autoFocus={true}
+                            className='form-control filter-textbox'
+                            type='text'
+                            value={this.state.categoryName}
+                            placeholder={localizeMessage('edit_category_modal.placeholder', 'Choose a category name')}
+                            clearable={true}
+                            onClear={this.handleClear}
+                            onChange={this.handleChange}
+                            maxlength={22}
+                        />
```
Yes, this came up in our weekly meeting as well. Will make sure to add.

devinbinnie commented on components/delete_category_modal.tsx:
```diff
@@ -0,0 +1,91 @@
+        return (
+            <Modal
+                dialogClassName='a11y__modal edit-category'
+                ref='modal'
```
Whoops, that's just leftovers from copying and pasting :P
Will remove.

devinbinnie commented on components/edit_category_modal.tsx:
```diff
@@ -0,0 +1,117 @@
+        return (
+            <Modal
+                dialogClassName='a11y__modal edit-category'
+                ref='modal'
```
Ditto on the copy/paste :P

hmhealey commented on components/sidebar/sidebar.tsx:
```diff
@@ -140,13 +177,17 @@ export default class Sidebar extends React.PureComponent<Props, State> {
-                <SidebarCategoryList handleOpenMoreDirectChannelsModal={this.handleOpenMoreDirectChannelsModal}/>
+                <SidebarCategoryList
+                    handleOpenMoreDirectChannelsModal={this.handleOpenMoreDirectChannelsModal}
+                    newCategoryIds={this.newCategoryIds}
```
`this.newCategoryIds` affects rendering, so it should go in `this.state.newCategoryIds`

hmhealey commented on components/sidebar/sidebar_category/sidebar_category.tsx:
```diff
@@ -97,26 +119,57 @@ export default class SidebarCategory extends React.PureComponent<Props> {
+        if (isNewCategory) {
+            newLabel = (
+                <div className='SidebarCategory_newLabel'>
+                    {localizeMessage('sidebar_left.sidebar_category.newLabel', 'new')}
```
This can use a `FormattedMessage` component since we don't need the translated string as a plain string

hmhealey commented on components/sidebar/sidebar_category/sidebar_category.tsx:
```diff
@@ -97,26 +119,57 @@ export default class SidebarCategory extends React.PureComponent<Props> {
+                >
+                    <i className='icon-hand-right'/>
+                    <span className='SidebarCategory_newDropBox-label'>
+                        {localizeMessage('sidebar_left.sidebar_category.newDropBoxLabel', 'Drag channels here...')}
```
Same here

hmhealey commented on sass/components/_categories.scss:
```diff
@@ -0,0 +1,140 @@
+@charset 'UTF-8';
```
Similar to the other PR, this CSS should live in the same folders as the accompanying components

hmhealey commented on components/delete_category_modal.tsx:
```diff
@@ -0,0 +1,91 @@
+
+    render() {
+        return (
+            <Modal
```
This component shares a lot of structure and CSS with the EditCategoryModal. How much of it could we factor out into a shared component instead of repeating the structure between the two modals? We could use that as the start of a new shared modal component for other things as well since we're going to need it for the What's New modal and such as well.

devinbinnie replied:
I do plan to create a modal that uses this style that we can use as a common component, but the plan was to tackle that after the main feature was completed.

hmhealey commented on components/sidebar/sidebar_category/sidebar_category.test.tsx:
```diff
@@ -40,6 +40,7 @@ describe('components/sidebar/sidebar_category', () => {
         setChannelRef: jest.fn(),
         getChannelRef: jest.fn(),
         handleOpenMoreDirectChannelsModal: jest.fn(),
+        isNewCategory: false,
```
Can we add some tests so that we know the right parts of the component are rendered when in this mode?

devinbinnie replied:
I haven't gotten to unit tests for this part yet, they are scheduled for later. But I can add this one now.

hmhealey commented on sass/layout/_sidebar-left.scss:
```diff
@@ -372,8 +427,11 @@
 
-        > div {
-            flex: 1 1 auto;
+        > div:not(.SidebarCategory_newLabel) {
```
Would it help to make this more specific than saying every div that's not the new category label? I don't know which parts of this are divs that need to get these styles.

devinbinnie replied:
It's mostly the ones for the menus and the + button for Direct Messages. I can be more specific about this once the PRs are merged together.
